test/preconditioner/batch_jacobi_kernels.cpp

```diff
@@ -0,0 +1,221 @@
+        // other device kernels
+        // // so that the block pointers are exactly the same for ref and device
+        // const int* block_pointers_generated_by_ref =
+        //     ref_block_jacobi_prec->get_const_block_pointers();
+        // const auto num_blocks_generated_by_ref =
+        //     ref_block_jacobi_prec->get_num_blocks();
+
+        // gko::array<int> block_pointers_for_device(
+        //     this->exec, block_pointers_generated_by_ref,
+        //     block_pointers_generated_by_ref + num_blocks_generated_by_ref +
+        //     1);
+
+        d_block_jacobi_prec =
+            BJ::build()
+                .with_max_block_size(max_blk_sz)
+                // .with_block_pointers(block_pointers_for_device)
+                .on(exec)
+                ->generate(d_mtx);
+    }
+
+    template <typename MatrixType>
+    gko::test::LinearSystem<MatrixType> setup_linsys_and_solver(
+        std::shared_ptr<const MatrixType> mat, const int num_rhs,
+        const real_type tol, const int max_iters)
+    {
+        auto executor = exec;
+        solve_lambda = [executor](const Settings settings,
+                                  const gko::batch::BatchLinOp* prec,
+                                  const Mtx* mtx, const MVec* b, MVec* x,
+                                  LogData& log_data) {
+            gko::kernels::EXEC_NAMESPACE::batch_bicgstab::apply<
+                typename Mtx::value_type>(executor, settings, mtx, prec, b, x,
+                                          log_data);
+        };
+        solver_settings = Settings{max_iters, tol,
+                                   gko::batch::stop::tolerance_type::relative};
+        precond_solver_factory =
+            solver_type::build()
+                .with_max_iterations(max_iters)
+                .with_tolerance(tol)
+                .with_tolerance_type(gko::batch::stop::tolerance_type::relative)
+                .with_preconditioner(
+                    precond_type::build().with_max_block_size(2u))
+                .on(exec);
+        return gko::test::generate_batch_linear_system(mat, num_rhs);
+    }
+
+    std::function<void(const Settings, const gko::batch::BatchLinOp*,
+                       const Mtx*, const MVec*, MVec*, LogData&)>
+        solve_lambda;
+    Settings solver_settings{};
+    std::shared_ptr<typename solver_type::Factory> precond_solver_factory;
+
+    const size_t nbatch = 3;
+    const int nrows = 300;
+    std::shared_ptr<Mtx> ref_mtx;
+    std::shared_ptr<Mtx> d_mtx;
+    std::unique_ptr<BMVec> ref_b;
+    std::unique_ptr<BMVec> d_b;
+    std::unique_ptr<BMVec> ref_x;
+    std::unique_ptr<BMVec> d_x;
+    const gko::uint32 max_blk_sz = 6u;
+    std::unique_ptr<BJ> ref_scalar_jacobi_prec;
+    std::unique_ptr<BJ> d_scalar_jacobi_prec;
+    std::unique_ptr<BJ> ref_block_jacobi_prec;
+    std::unique_ptr<BJ> d_block_jacobi_prec;
+};
+
+
+TEST_F(BatchJacobi, BatchBlockJacobiGenerationIsEquivalentToRef)
```

Comment:
miss the scalar

Reply:
Scalar Jacobi generation is not done explicitly but directly within the solver call, so it is a bit difficult to test that. I added a solver test here.